### PR TITLE
Add single selection mode to SelectTree

### PR DESCRIPTION
**Pull request by nathanss**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Add single selection mode to SelectTree.

https://user-images.githubusercontent.com/13437655/236469085-deeb839b-0507-461b-a3ae-4884ff33a5b0.mov

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Launch storybook.
2. Go to http://localhost:6007/?path=/story/woocommerce-admin-experimental-selecttreecontrol--single-select-tree
3. Test functionality with mouse
4. Test functionality by selecting from the list with keyboard.
5. Test functionality by manually typing a correct value on the input.

### CODE REVIEW

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx:1]: Whoa, whoa, whoa... why are we disabling `@typescript-eslint/no-explicit-any` at the file level?! We shouldn't be doing this. It greatly diminishes the type checking.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx [hunk 1]
```diff
@@ -48,6 +49,68 @@ export const SelectTree = function SelectTree( {
...
+		}
+	};
+
+	const focusOnInput = () => {
+		(
+			document.querySelector( `#${ props.id }-input` ) as HTMLInputElement
+		 )?.focus();
+	};
+
+	const inputProps: any = {
```
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx]: We shouldn't use `any` -- use a specific type.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx [hunk 2]
```diff
@@ -48,6 +49,68 @@ export const SelectTree = function SelectTree( {
...
+		'aria-autocomplete': 'list',
+		'aria-controls': `${ props.id }-menu`,
+		autoComplete: 'off',
+		onFocus: () => {
+			if ( ! isOpen ) {
+				setIsOpen( true );
+			}
+			setIsFocused( true );
+		},
+		onBlur: ( event: any ) => {
```
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx]: We shouldn't use `any` -- use a specific type.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx [hunk 3]
```diff
@@ -48,6 +49,68 @@ export const SelectTree = function SelectTree( {
...
+				! document
+					.querySelector( '.' + menuInstanceId )
+					?.contains( event.relatedTarget )
+			) {
+				setIsOpen( false );
+				recalculateInputValue();
+			}
+			setIsFocused( false );
+		},
+		onKeyDown: ( event: any ) => {
```
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx]: We shouldn't use `any` -- use a specific type.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx [hunk 4]
```diff
@@ -48,6 +49,68 @@ export const SelectTree = function SelectTree( {
...
+						`.${ menuInstanceId } input, .${ menuInstanceId } button`
+					) as HTMLInputElement | HTMLButtonElement
+				 )?.focus();
+			}
+			if ( event.key === 'Tab' ) {
+				setIsOpen( false );
+				recalculateInputValue();
+			}
+		},
+		onChange: ( event: any ) =>
```
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx]: We shouldn't use `any` -- use a specific type.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx [hunk 5]
```diff
@@ -68,79 +131,73 @@ export const SelectTree = function SelectTree( {
...
+								'woocommerce-experimental-select-control__combo-box-wrapper',
+							role: 'combobox',
+							'aria-expanded': isOpen,
+							'aria-haspopup': 'tree',
+							'aria-labelledby': `${ props.id }-label`,
+							'aria-owns': `${ props.id }-menu`,
+						} }
+						inputProps={ {
+							...inputProps,
+							onChange: ( event: any ) =>
```
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx]: We shouldn't use `any` -- use a specific type.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx [hunk 6]
```diff
@@ -48,6 +49,68 @@ export const SelectTree = function SelectTree( {
...
+			}
+			setIsFocused( true );
+		},
+		onBlur: ( event: any ) => {
+			// if blurring to an element inside the dropdown, don't close it
+			if (
+				isOpen &&
+				! document
+					.querySelector( '.' + menuInstanceId )
+					?.contains( event.relatedTarget )
```
Comment (mattsherman) [inline packages/js/components/src/experimental-select-tree-control/select-tree.tsx]: You could extract this to a function and do something like `! isInsideDropdown( event.relatedTarget )`... if you did that, you don't even need the comment... the code self-documents.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Overall, works as expected. Nice job.

There is a lot of use of `any` in the code. This should be avoided as it really weakens our type-checking.

**Review by mattsherman — CHANGES_REQUESTED** [review]
Overall, works as expected. Nice job.

There is a lot of use of `any` in the code. This should be avoided as it really weakens our type-checking.

File: packages/js/components/src/experimental-select-tree-control/select-tree.tsx
Comment (nathanss) [reply packages/js/components/src/experimental-select-tree-control/select-tree.tsx:1]: My bad. This was by mistake, I disable it sometimes just to speed things up while I'm testing things, but it was never intended to reach production 😛 Thanks for catching it.

**Review by joshuatf — COMMENTED** [review]
Just a heads up that we're having a discussion around drop-down behavior tomorrow with @jarekmorawski.

I think we need to DRY up our use of `SelectTree` and `SelectControl` as there is a lot of duplication and fix divergence in behavior as best we can.  If you can attend @nathanss, your input would be valuable.

**Comment by mattsherman** [comment]
@nathanss Can you please resolve the conflict in the following file before I re-review it?

- `packages/js/components/src/experimental-select-tree-control/select-tree.tsx`

**Review by mattsherman — APPROVED** [review]
Nice iteration on this, @nathanss. 🚢

**Pull request merged**